tests/integration/ha_tests/application-charm/src/charm.py

```diff
@@ -128,33 +141,26 @@ def _start_continuous_writes(self, starting_number: int) -> None:
         )
 
         # Store the continuous writes process ID to stop the process later.
-        self._stored.continuous_writes_pid = popen.pid
+        self.app_peer_data[PROC_PID_KEY] = str(popen.pid)
 
     def _stop_continuous_writes(self) -> Optional[int]:
         """Stops continuous writes to PostgreSQL and returns the last written value."""
-        if self._stored.continuous_writes_pid is None:
+        if not self.app_peer_data.get(PROC_PID_KEY):
             return None
 
         # Stop the process.
-        proc = subprocess.Popen(["pkill", "--signal", "SIGKILL", "-f", "src/continuous_writes.py"])
-
-        # Wait for process to be killed.
-        proc.communicate()
+        os.kill(int(self.app_peer_data[PROC_PID_KEY]), signal.SIGTERM)
 
-        self._stored.continuous_writes_pid = None
+        del self.app_peer_data[PROC_PID_KEY]
 
         # Return the max written value (or -1 if it was not possible to get that value).
         try:
-            for attempt in Retrying(stop=stop_after_delay(60), wait=wait_fixed(3)):
+            for attempt in Retrying(stop=stop_after_delay(60), wait=wait_fixed(5)):
                 with attempt:
-                    with psycopg2.connect(
-                        self._connection_string
-                    ) as connection, connection.cursor() as cursor:
-                        cursor.execute("SELECT MAX(number) FROM continuous_writes;")
-                        last_written_value = int(cursor.fetchone()[0])
-                    connection.close()
+                    with open(LAST_WRITTEN_FILE, "r") as fd:
+                        last_written_value = int(fd.read())
         except RetryError as e:
-            logger.exception(e)
+            logger.exception("Unable to query the database", exc_info=e)
```

Comment:
Shouldn't we change this message to something related to reading the file instead of querying the database?